File: windows/WallpaperInfoApp/WallpaperInfoApp/WallpaperServiceInfo.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WallpaperInfoApp {

	class WallpaperServiceInfo {

		private WallpaperServiceInfo() {
			var wpref = PlatformPreference.getPreferences(this);
            _sourceRootPath = wpref.root_path;
            _customThemeInfo._root = wpref.custom_root;
            _customThemeInfo._allow = wpref.custom_allow;
            _customThemeInfo._filter = wpref.custom_filter;
            // must update custom info before it updates themeinfo
            ThemeInfo.setCustomConfig(_customThemeInfo._root, _customThemeInfo._allow, _customThemeInfo._filter);
            _themeInfo = ThemeInfo.getThemeInfo(wpref.theme); 
            _pause = wpref.pause;
            _interval = wpref.interval;
			if (_interval < _minInterval || _interval > _maxInterval) {
				_interval = _defaultInterval;
			}
			if (_sourceRootPath == null || !BPUtil.fileExists(_sourceRootPath)) {
				_sourceRootPath = ImageFileManager.getDefaultSourceRootPath();
			}
		}

		private static WallpaperServiceInfo _sharedWallpaperServiceInfo;
		public static WallpaperServiceInfo getInstance() {
			if (_sharedWallpaperServiceInfo == null) {
				_sharedWallpaperServiceInfo = new WallpaperServiceInfo();
            }
			return _sharedWallpaperServiceInfo;
        }

		// persistent property setter/getter
		 public String getSourceRootPath() {
			return _sourceRootPath;
		}
		public void setSourceRootPath(String path) {
			_sourceRootPath = path;
			PlatformPreference.setPreference("root_path", path);
		}
		public Theme getTheme() { return _themeInfo._theme; }
		public ThemeInfo getThemeInfo() { return _themeInfo; }
		public void setThemeInfo(ThemeInfo themeInfo) {
			_themeInfo = themeInfo;
			PlatformPreference.setPreference("theme", themeInfo._theme.intValue().ToString());
		}
		public void setCustomThemeInfo(ThemeInfo customThemeInfo) {
			_customThemeInfo = customThemeInfo;
			PlatformPreference.setPreference("custom_root", customThemeInfo._root);
			PlatformPreference.setPreference("custom_allow", customThemeInfo._allow);
			PlatformPreference.setPreference("custom_filter", customThemeInfo._filter);
		}
		public String getCustomConfigString() {
			return _customThemeInfo._root + ";" + _customThemeInfo._allow + ";" + _customThemeInfo._filter;
		}
		public bool getPause() {
			return _pause;
		}
		public void setPause(bool pause) {
    		_pause = pause;
			PlatformPreference.setPreference("pause", pause.ToString());
		}
		public int getInterval() {
			return _interval;
		}
		public void setInterval(int interval) {
			_interval = interval;
			PlatformPreference.setPreference("interval", interval.ToString());
		}

		// runtime property setter/getter
		public WPath getcurrentWPath() {
			return _currentWPath;
		}
		public void setcurrentWPath(WPath wpath) {
			_currentWPath = wpath;
		}

		public Bitmap getThumbnail() {
			return _thumbnail;
		}
		public void setThumbnail(Bitmap thumbnail) {
		if (thumbnail == null && _thumbnail != null) {
			_thumbnail.Dispose();
		}
			_thumbnail = thumbnail;
		}

		public WLinkedHashMap<String, String> getLastUsedPaths() {
			return _lastUsedPaths;
		}
		public void setLastUsedPaths(WLinkedHashMap<String, String> lastUsedPaths) {
			_lastUsedPaths = lastUsedPaths;
		}

		public WallpaperService getWallpaperService() { return _wallpaperService; }

		// persistent property
		private String _sourceRootPath; // ImageFileManager.getDefaultRootPath()
		private ThemeInfo _themeInfo = ThemeInfo.getThemeInfo(Theme.default1);
		private ThemeInfo _customThemeInfo = new ThemeInfo(Theme.custom,
				ThemeInfo._default_custom_root,
				ThemeInfo._default_custom_allow,
				ThemeInfo._default_custom_filter);
		private bool _pause; // false
		private int _interval; // _defaultInterval // in seconds

		// runtime property
		private WPath _currentWPath = null;
		private Bitmap _thumbnail = null;
		private WLinkedHashMap<String, String> _lastUsedPaths = new WLinkedHashMap<String, String>();

		// Platform Specifc runtime property
		private WallpaperService _wallpaperService = new WallpaperService();

		//public value from resource
		public static int _defaultInterval = 5; // 15 will not be used. just placeholder.
		public static int _minInterval = 5; // 5 will not be used. just placeholder.
		public static int _maxInterval = 30; // 30 will not be used. just placeholder.
	}
}

File: apps/desktop-macosx/WallpaperInfoApp/WallpaperInfoApp/WallpaperServiceInfo.swift
//
//  WallpaperServiceInfo.swift
//  WallpaperInfoApp
//
//  Created by Brad Park on 4/11/20.
//  Copyright © 2020 Brad Park. All rights reserved.
//

import Foundation
import Cocoa

class WallpaperServiceInfo {

    private init() {
        let pref = PlatformPreference.getPreferences()
        _sourceRootPath = pref.root_path
        _customThemeInfo._root = pref.custom_root
        _customThemeInfo._allow = pref.custom_allow
        _customThemeInfo._filter = pref.custom_filter
        // must update custom info before it updates themeinfo
        let _ = ThemeInfo.setCustomConfig(root: _customThemeInfo._root, allow: _customThemeInfo._allow, filter: _customThemeInfo._filter)
        _themeInfo = ThemeInfo.getThemeInfo(theme: pref.theme)
        _pause = pref.pause
        _interval = pref.interval
        if (_interval < WallpaperServiceInfo._minInterval || _interval > WallpaperServiceInfo._maxInterval) {
            _interval = WallpaperServiceInfo._defaultInterval;
        }
        if (!BPUtil.fileExists(_sourceRootPath)) {
            _sourceRootPath = ImageFileManager.getDefaultSourceRootPath();
        }
    }

    private static var sharedWallpaperServiceInfo: WallpaperServiceInfo = {
        let instance = WallpaperServiceInfo()
        return instance
    }()
    
    class func getInstance() -> WallpaperServiceInfo {
        return sharedWallpaperServiceInfo
    }
    
    // persistent property setter/getter
    func getSourceRootPath() -> String { return _sourceRootPath }
    func setSourceRootPath(_ path: String) {
        _sourceRootPath = path;
		PlatformPreference.setPreference(key:"root_path", value:path)
    }
	
	func getTheme() -> Theme { return _themeInfo._theme }
	func getThemeInfo() -> ThemeInfo { return _themeInfo }
	func setThemeInfo(_ themeInfo: ThemeInfo) {
		_themeInfo = themeInfo
		PlatformPreference.setPreference(key:"theme", value:String(themeInfo._theme.intValue))
	}
	func setCustomThemeInfo(_ customThemeInfo: ThemeInfo) {
		_customThemeInfo = customThemeInfo
		PlatformPreference.setPreference(key:"custom_root", value:customThemeInfo._root)
		PlatformPreference.setPreference(key:"custom_allow", value:customThemeInfo._allow)
		PlatformPreference.setPreference(key:"custom_filter", value:customThemeInfo._filter)
    }
	func getCustomConfigString() -> String {
		return _customThemeInfo._root + ";" + _customThemeInfo._allow + ";" + _customThemeInfo._filter
	}
    
    func getPause() -> Bool { return _pause }
    func setPause(_ pause: Bool) {
        _pause = pause;
		PlatformPreference.setPreference(key:"pause", value:String(pause))
    }
    
    func getInterval() -> Int { return _interval }
    func setInterval(_ interval: Int) {
        _interval = interval;
		PlatformPreference.setPreference(key:"interval", value:String(interval))
    }
    
    // runtime property setter/getter
    func getWallpaperService() -> WallpaperService { return _wallpaperService }
    
    func getcurrentWPath() -> WPath? { return _currentWPath }
    func setcurrentWPath(_ path: WPath?) { _currentWPath = path }
    
    func getThumbnail() -> NSImage? { return _thumbnail }
    func setThumbnail(_ thumbnail: NSImage?) {
        if (thumbnail == nil && _thumbnail != nil) {
            // TODO find a way to release resource
            // _thumbnail.recyle()
        }
        _thumbnail = thumbnail
    }

    func getLastUsedPaths() -> WLinkedHashMap<String, String> { return _lastUsedPaths }
    func setLastUsedPaths(_ lastUsedPaths: WLinkedHashMap<String, String>) { _lastUsedPaths = lastUsedPaths }

    func getMode() -> ServiceMode { return _mode }
	func setMode(_ mode:ServiceMode) { _mode = mode }

	func getSaver() -> Bool { return _saver }
	func setSaver(_ saver:Bool) { _saver = saver }
	
	func getSaverTime() -> Int { return _saverTime }
	func setSaverTime(_ saverTime:Int) { _saverTime = saverTime }
	
    // persistent property
    private var _sourceRootPath:String // ImageFileManager.getDefaultSourceRootPath()
	private var _themeInfo:ThemeInfo = ThemeInfo.getThemeInfo(theme: .default1);
	private var _customThemeInfo:ThemeInfo = ThemeInfo(theme: .custom,
																		 root: ThemeInfo._default_custom_root,
																		 allow: ThemeInfo._default_custom_allow,
																		 filter: ThemeInfo._default_custom_filter)
    private var _pause:Bool // false
    private var _interval:Int // _defaultInterval // in seconds

    // runtime property
    private var _currentWPath:WPath? = nil
    private var _thumbnail:NSImage? = nil;
    private var _lastUsedPaths:WLinkedHashMap<String, String> = WLinkedHashMap<String, String>()
	private var _mode:ServiceMode = .wallpaper
	private var _saver:Bool = false;
	private var _saverTime:Int = 5;
	
	// Platform Specifc runtime property
	lazy private var _wallpaperService:WallpaperService = WallpaperService()
    
    //public value from resource
    public static let _defaultInterval:Int = 5; // 15 will not be used. just placeholder.
    public static let _minInterval:Int = 5; // 5 will not be used. just placeholder.
    public static let _maxInterval:Int = 30; // 30 will not be used. just placeholder.
}

enum ServiceMode:Int {
	case wallpaper = 0
	case slideshow
	case cast

	var intValue: Int {
		return rawValue
	}
}
